File: IQ/Classes/clsBands.cs
public class clsBand
{
    public long min;
    public long max;
    public int Survivors; //The number of values falling in this band

    public clsBand(long min, long max, int Survivors)
    {
        this.min = min;
        this.max = max;
        this.Survivors = Survivors;
    }

    public bool contains(long numericValue)
    {
        bool returnValue = false;
        returnValue = false;
        if (numericValue >= this.min & numericValue <= this.max)
        {
            returnValue = true;
        }
        return returnValue;
    }

    public void Stretch()
    {

        //do some rounding and overlapping here on the Max/Min

        this.min = System.Convert.ToInt64(System.Convert.ToInt32(Math.Ceiling((double)this.min / 1000)) * 1000);
        this.max = System.Convert.ToInt64(System.Convert.ToInt32(Math.Floor((double)this.max / 1000)) * 1000);


    }

    /// <summary>
    /// Compares the Min and Max of this band to those values specified for the LE and GE filters (in the dictionary, for the field provided - to determin with this band is the currently selected one
    /// </summary>
    /// <param name="fld"></param>
    /// <param name="filters"></param>
    /// <returns></returns>
    /// <remarks></remarks>
    public bool isSelected(clsField fld, Dictionary<clsField, Dictionary<clsFilter, List<long>>> filters)
    {
        bool returnValue = false;

        returnValue = false;

        if (filters.ContainsKey(fld))
        {
            clsFilter ge = iq.i_Filters_Code("GE");
            clsFilter le = iq.i_Filters_Code("LE");

            Dictionary with_1 = filters(fld);

            if (with_1.ContainsKey(ge))
            {
                if (with_1.ContainsKey(le))
                {
                    object mi = with_1.Item(ge).IndexOf(this.min);
                    if ((int)mi == -1)
                    {
                        return false;
                    }
                    if (with_1.Item(le)[mi] == this.max)
                    {
                        return true;
                    }
                    //If .Item(ge).First() = Me.min And .Item(le).First() = Me.max Then Return True 'ML - have guessed here that there will only be one min and one max, will need to enforce that this is the case - TODO
                }
            }
        }
        return returnValue;
    }

    public bool Equals(clsBand obj)
    {
        return max == (ulong)obj.max && min == obj.min;
    }

    public override int GetHashCode()
    {
        return max.GetHashCode() + min.GetHashCode();
    }

}

File: IQ/Classes/clsBands.vb

Public Class clsBand
    Public min As Int64
    Public max As Int64
    Public Survivors As Integer 'The number of values falling in this band

    Public Sub New(min As Int64, max As Int64, Survivors As Integer)
        Me.min = min
        Me.max = max
        Me.Survivors = Survivors
    End Sub

    Public Function contains(numericValue As Int64) As Boolean
        contains = False
        If numericValue >= Me.min And numericValue <= Me.max Then
            contains = True
        End If
    End Function

    Public Sub Stretch()

        'do some rounding and overlapping here on the Max/Min

        Me.min = CType(Math.Ceiling(Me.min / 1000) * 1000, Int64)
        Me.max = CType(Math.Floor(Me.max / 1000) * 1000, Int64)


    End Sub

    ''' <summary>
    ''' Compares the Min and Max of this band to those values specified for the LE and GE filters (in the dictionary, for the field provided - to determin with this band is the currently selected one
    ''' </summary>
    ''' <param name="fld"></param>
    ''' <param name="filters"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function isSelected(fld As clsField, filters As Dictionary(Of clsField, Dictionary(Of clsFilter, List(Of Int64)))) As Boolean

        isSelected = False

        If filters.ContainsKey(fld) Then
            Dim ge As clsFilter = iq.i_Filters_Code("GE")
            Dim le As clsFilter = iq.i_Filters_Code("LE")

            With filters(fld)

                If .ContainsKey(ge) Then
                    If .ContainsKey(le) Then
                        Dim mi = .Item(ge).IndexOf(Me.min)
                        If mi = -1 Then Return False
                        If .Item(le)(mi) = Me.max Then Return True
                        'If .Item(ge).First() = Me.min And .Item(le).First() = Me.max Then Return True 'ML - have guessed here that there will only be one min and one max, will need to enforce that this is the case - TODO
                    End If
                End If
            End With
        End If
    End Function

    Public Overloads Function Equals(obj As clsBand) As Boolean
        Return max = CULng(obj.max) AndAlso min = obj.min
    End Function

    Public Overrides Function GetHashCode() As Integer
        Return max.GetHashCode() + min.GetHashCode()
    End Function

End Class

